api/v1beta2/ibmvpccluster_types.go

```diff
@@ -98,26 +117,136 @@ type AdditionalListenerSpec struct {
+// kubebuilder:validation:XValidation:rule="self.dely > self.timeout",message="health monitor's delay must be greater than the timeout"
+type HealthMonitorSpec struct {
+	// delay defines the seconds to wait between health checks.
+	// +kubebuilder:validation:Minimum=2
+	// +kubebuilder:validation:Maximum=60
+	// +required
+	Delay int64 `json:"delay"`
+
+	// retries defines the max retries for health check.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=10
+	// +required
+	Retries int64 `json:"retries"`
+
+	// port defines the port to perform health monitoring on.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	// +optional
+	Port *int64 `json:"port,omitempty"`
+
+	// timeout defines the seconds to wait for a health check response.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=59
+	// +required
+	Timeout int64 `json:"timeout"`
+
+	// type defines the protocol used for health checks.
+	// +required
+	Type VPCLoadBalancerBackendPoolHealthMonitorType `json:"type"`
+
+	// urlPath defines the URL to use for health monitoring.
+	// +kubebuilder:validation:Pattern=`^\/(([a-zA-Z0-9-._~!$&'()*+,;=:@]|%[a-fA-F0-9]{2})+(\/([a-zA-Z0-9-._~!$&'()*+,;=:@]|%[a-fA-F0-9]{2})*)*)?(\\?([a-zA-Z0-9-._~!$&'()*+,;=:@\/?]|%[a-fA-F0-9]{2})*)?$`
+	// +optional
+	URLPath *string `json:"urlPath,omitempty"`
+}
+
+// ImageSpec defines the desired state of the VPC Custom Image resources for the cluster.
+// +kubebuilder:validation:XValidation:rule="(!has(self.cosInstance) && !has(self.cosBucket) && !has(self.cosObject)) || (has(self.cosInstance) && has(self.cosBucket) && has(self.cosObject))",message="if any of cosInstance, cosBucket, or cosObject are specified, all must be specified"
+// +kubebuilder:validation:XValidation:rule="has(self.name) || has(self.crn) || has(self.catalogOffering) || (has(self.cosInstance) && has(self.cosBucket) && has(self.cosObject))",message="an existing image name or crn or a catalog offering must be provided, or to create a new image the cos resources must be provided, with or without a name"
+// +kubebuilder:validation:XValidation:rule="!has(self.catalogOffering) || (has(self.catalogOffering) && (has(self.catalogOffering.catalogCRN) || has(self.catalogOffering.versionCRN)) && !(has(self.catalogOffering.catalogCRN) && has(self.catalogOffering.versionCRN)))",message="a provided catalogOffering must have either a catalogCRN or a versionCRN, but not both"
+type ImageSpec struct {
+	// name is the name of the desired VPC Custom Image.
+	// +kubebuilder:validation:MinLength:=1
+	// +kubebuilder:validation:MaxLength:=63
+	// +kubebuilder:validation:Pattern='/^([a-z]|[a-z][-a-z0-9]*[a-z0-9])$/'
+	// +optional
+	Name *string `json:"name,omitempty"`
+
+	// catalogOffering defines an existing Catalog Offering Image.
+	// +optional
+	CatalogOffering *CatalogOffering `json:"catalogOffering,omitempty"`
```

Comment:
`CatalogOffering`'s spec looks like to identify generic ibm cloud catalog offering, how are you planning to use that to identify the catalog image?

Reply:
Using a Catalog Offering for a VSI "image" requires either a Catalog Offering CRN or Catalog Offering Version CRN (with an optional Catalog Offering Plan CRN for either) per the IBM Cloud VPC API's.

Supplying the `catalogOffering` field for the `ImageSpec`, with one of the offeringCRN or versionCRN, with result in the VSI API calls to be made using an `InstancePrototypeInstanceByCatalogOffering` instead of the `...ByImage`  for the API calls to create a VSI.

Reply:
I dropped the Catalog Offering since we may not have permission to validate the Image as an Offering. We will have to rely on results/errors during Machine creation.